### ISSUE TITLE
LocalFinalVariableName: false negative on pattern variable

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#10971]
https://checkstyle.org/config_naming.html#LocalFinalVariableName

````
$ cat TestClass.java
public class Test {
    public void run(String... arguments) {
        final Object o = 45;
        if (o instanceof final String BAD) {
            System.out.println(BAD);
        }
    }
}

$ cat TestConfig.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
<module name="LocalFinalVariableName"/>
<module name="LocalVariableName"/>
    </module>
</module>

$ java -jar checkstyle-9.0.1-all.jar -c TestConfig.xml TestClass.java
Starting audit...
Audit done.
````

I am expecting a violation on line 5 since `BAD` is a local, final variable and it doesn't start with a lower case as is default required.

```
public class Test {
    public void run(String... arguments) {
        final Object o = 45;
        if (o instanceof final String BAD) {
            o = 34;  // javac: Test.java:5: error: cannot assign a value to final variable o
            System.out.println(BAD);
        }
    }
}
```

**Comment by rnveach** [comment]
I guess this may require more discussion since we have PatternVariableName ( https://checkstyle.org/config_naming.html#PatternVariableName ) and we don't want to overlap functionality. Not sure if it makes sense to add a new check for final patterns. At the least, we should update documentation to point to the other check for patterns.

**Comment by romani** [comment]
We need to update documentation to state clearly that this Check validate variables with final modifier.

There are lot of cases when variable becomes final even without modifier.

**Comment by rnveach** [comment]
> validate variables with final modifier.

But this code has a final variable on it, so it needs to be more explicit than that. Also, since we have 2 checks for local variable names, you didn't cover having 2 checks for pattern varaible names.

**Comment by ShivanshuPandey007** [comment]
May I fix this issue but I am a beginner so please correct me if i am wrong . what i get is you want:
>update documentation to state clearly that this Check validate variables with final modifier and variable
>change pattern variable names accordingly

**Comment by romani** [comment]
issue is not approved yet, better to avoid it.

--------

I think LocalVariableName should not  have validation on this code as we have PatternVariableName .
but LocalFinalVariableName can validate this , as "final" is explicit in such code, as above example shows - it is verified by compiler that value cannot be changed.

@pbludov , @strkkk , @nmancus1  - please share what is expected behavior.

**Comment by strkkk** [comment]
I think LocalFinalVariableNameCheck should handle such case and issue is valid

**Comment by nmancus1** [comment]
For usual local variables, we have `LocalVariableName` and ` LocalFinalVariableName`. `LocalVariableName` does not check variables declared as `final`.

If we want to check pattern variables declared as `final` differently, then perhaps analogous idea with pattern variables should hold:

```
public class Test {
    public void run(String... arguments) {
        final Object BAD1 = 45; // violation by LocalFinalVariableName only
        Object BAD2 = 42;  // violation by LocalVariableName only
        Object o = 42;
        if (o instanceof final String BAD3) { // should be FinalPatternVariableName?
            System.out.println(BAD3);
        }
        if (o instanceof String BAD4) { // should be PatternVariableName only
            System.out.println(BAD4);
        }
    }
}

```

Current behavior:

```
➜  company cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
          "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
          "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <property name="charset" value="UTF-8"/>

    <module name="TreeWalker">
        <module name="LocalFinalVariableName"/>
        <module name="LocalVariableName"/>
        <module name="PatternVariableName"/>
    </module>
</module>  

➜  java -jar checkstyle-9.2-all.jar -c config.xml Test.java
Starting audit...
[ERROR] Test.java:3:22: Name 'BAD1' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalFinalVariableName]
[ERROR] Test.java:4:16: Name 'BAD2' must match pattern '^[a-z][a-zA-Z0-9]*$'. [LocalVariableName]
[ERROR] Test.java:6:39: Name 'BAD3' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
[ERROR] Test.java:9:33: Name 'BAD4' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
Audit done.
Checkstyle ends with 4 errors.
```

So, any fix for this issue should also exclude final pattern variables from being checked in `PatternVariableNameCheck`.

Here is another idea: now that the statement from [here](https://docs.oracle.com/en/java/javase/14/docs/specs/patterns-instanceof-jls.html#jls-6.3.1) "Pattern variables are implicitly final by design" is not accurate any longer, would it make sense to just add pattern variable support to `LocalVariableName` and `LocalFinalVariableName` and do away with `PatternVariableName`?

I think the two best options are either:
(1) Make a new check, `FinalPatternVariableNameCheck`, or
(2) Add support to `LocalVariableName` and `LocalFinalVariableName` and do away with `PatternVariableName`

Pattern matching/ variable extension is imminent, as seen in [JEP406](https://openjdk.java.net/jeps/406) and [JEP405](https://openjdk.java.net/jeps/405), we may regret grouping pattern variables in with usual local variables later.

**Comment by pbludov** [comment]
Why do we even have `LocalFinalVariableName`? Is there any team whose coding style requires different names for variables and for final variables?

Perhaps for someone
```java
   int i = 0;
   ...
   ++i;
```
is OK, but
```java
   final int i = 0;
   ...
   return i;
```
is not OK.

If so, it is unlikely that such convention suitable for pattern variables.
I think we need only one check for both pattern variables and final pattern variables.
It will be easy to split this check  if anyone requests it.

**Comment by nmancus1** [comment]
>  Is there any team whose coding style requires different names for variables and for final variables?

This is a good question; I could not find a use case for this.

I agree with @pbludov, let's let user request this; until then, we check any pattern variable with `PatternVariableNameCheck`. If user has immediate need for differentiation, they can use XPathSuppression and two `PatternVariableNameCheck` modules while we split the check.

**Comment by rnveach** [comment]
> Why do we even have LocalFinalVariableName? Is there any team whose coding style requires different names for variables and for final variables?

1) It is up to each team on their style. I don't know any, but we also find time and time again the simpler the check, the better.
2) It is because no other check validates it. If it is not validated, then there is no enforcement of a style and users can put in whatever they want.

> let user request this

I am requesting a check to validate this. I have no big preference if it is a new check or update existing one.

**Comment by nrmancuso** [comment]
We don't need a new check for this, we can easily achieve this behavior with PatternVariableCheck and a SuppressionXpathSingleFilter module:

```
➜  src cat Test.java 
public class Test {

    void m1(Object o) {
        // ok below, not final
        if (o instanceof String BAD) {
            System.out.println(BAD);
        }
    }

    void m2(Object o) {
        // violation below, final
        if (o instanceof final String BAD) {
            System.out.println(BAD);
        }
    }
}

➜  src cat config.xml 
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE module PUBLIC "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="PatternVariableName"/>
    <module name="SuppressionXpathSingleFilter">
      <property name="checks" value="PatternVariableName"/>
      <property name="query" value="//PATTERN_VARIABLE_DEF[not(./MODIFIERS/FINAL)]/IDENT"/>
    </module>
  </module>
</module>

➜  src javac Test.java
                                             
➜  src java -jar checkstyle-10.13.0-all.jar -c config.xml Test.java
Starting audit...
[ERROR] /home/nick/IdeaProjects/tester/src/Test.java:12:39: Name 'BAD' must match pattern '^[a-z][a-zA-Z0-9]*$'. [PatternVariableName]
Audit done.
Checkstyle ends with 1 errors.
```

@rnveach if you agree, let's close this issue

**Comment by rnveach** [comment]
To be honest, I don't really support xpath in these type of scenarios.

https://github.com/checkstyle/checkstyle/issues/10971#issuecomment-974867505
> Not sure if it makes sense to add a new check for final patterns. At the least, we should update documentation to point to the other check for patterns.

**Comment by mahfouz72** [comment]
@rnveach @nrmancuso 
what about this solution ?

- add support for patterm variable in `localVariableName`
- completly remove `patternVariableName`
- support pattern variable in `localFinalVariableName`

please explain to me why we need seprate checks for naming local and pattern variables 

I think the naming convention is same for both even the concept of underscore (unnamed) variable is in both

if we want for some reason to violate locals only we can make a property in localVariableName called ignore pattern variables

**Comment by mahfouz72** [comment]
> Pattern variables will continue to become more used and appear in more constructs as Java introduces further new language features
Yes but that fact that they always will be just as any normal local variables will never changed right? and the naming conventions will not change

an example for this is record pattern we reused the `PATTERN_VARIABLE_DEF` token in it. so in terms of tree and checks the pattern variables in record is just as the pattern variables with enhanced instanceof


> Have we considered supporting record patterns in this check? Will it make this check more complicated?

I am not sure whick check do you refer to by "this check" but I assume it is localVariableName
no I don't see any complications here it is just a `PATTERN_VARIABLE_DEF` so  we make the check stop on this token ( it includes pattern variables in record pattern or enhanced instanceof) and `VARIABLE_DEF` for normal local variables

**Comment by nrmancuso** [comment]
We typically don't "remove" checks (especially newer ones), so this idea is out for sure.

It is good to have separate checks for pattern and local variables due to the scope of the variable; I might be ok with `p` being a name for a pattern variable, but a local variable should be a bit more descriptive since it's scope is larger.

I think that it does not make sense to add support for this to FinalLocalVariableName, since this does not follow our established pattern (we have PatternVariableName, so we should have FinalPatternVariableName if anything).

My ideas to solve this issue, in order of preference:
1. Use module + suppression: this is likely to be rarely used, so it is better for us to have this simple suppression that works 100%
2. New check
3. Add some property to PatternVariableName
4. Add support to LocalFinalVariableName

I am easily good with 1 or 2. @mahfouz72  @rnveach please render a decision on this, let's get this issue approved.

**Comment by mahfouz72** [comment]
I am ok with the first 3 options,for me 3 1 2 in order 
the property is validateFinalPatternVariables

**Comment by rnveach** [comment]
My original stance at https://github.com/checkstyle/checkstyle/issues/10971#issuecomment-2146443589 was that if we weren't going to update any checks, then we should at least update documentation in local variable checks. When I first created this issue, they looked like local variables to me. This is why I was not closing this issue.

> please explain to me why we need seprate checks for naming local and pattern variables
>  It is highly likely that no one will actually care about different naming for final pattern variables

This is what confused me and why I think we should update documentation if we are not doing check updates.

I think this question can really only be answered by your best practices scan. IMO, I think we should tie check update(s) in this issue to this part of your work. I didn't bring this up as part of some practice really, it was just seemed like something we were missing and then became slightly confusing.

IMO, if you don't find anything from a best practice scan, then this issue will only be documentation update (and suppression method). @nrmancuso 

> To be honest, I don't really support xpath in these type of scenarios.
> What are “these types of scenarios”? 

This is just my opinion. It is basically when you are using an xpath to suppress part of the check, especially if we say this is how to handle the check officially. In a way, it seems like we are agreeing to the issue but not agreeing to an official fix. In addition, I see xpath as being generally slow since it has to do all the computations and tracing to find the token(s) being specified. I am not sure if we have done any numbers on it's speed, but we are limited by the library itself. Since I have the knowledge, I think I would generally create my own check over using xpath.

Also your example only showed how to suppress one form of patterns. To validate both, you would have to add 2 instances of `PatternVariableName` and 2 `SuppressionXpathSingleFilter`, both with different IDs, to ensure everything is validated.

> My ideas to solve this issue, in order of preference:

I go back to my previous point of best practices. If that fails, we basically can provide point 1, but also include the previously mentioned documentation update that the local variable checks do not cover patterns, and to go to the pattern version of the check.

**Comment by nrmancuso** [comment]
@mahfouz72 please see if you can find some advice about naming final local variables differently at all, I have never heard of this practice. 

> I go back to my previous point of best practices. If that fails, we basically can provide point 1, but also include the previously mentioned documentation update that the local variable checks do not cover patterns, and to go to the pattern version of the check.

I am good to do this if we can’t find any good cause to support naming final pattern variables differently

**Comment by nrmancuso** [comment]
Ok, we have sat on this issue for a week. I am good to update docs with details of unnamed variables and an example with the suppression that I have shown above.

@rnveach please approve if you are good
